File: leetcode/Arrays/leetcode-0048/48.go
package arrays

func rotate(matrix [][]int) {
	if matrix == nil || len(matrix) == 0 || len(matrix[0]) == 0 {
		return
	}
	n := len(matrix)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if i+j >= n-1 {
				continue
			}
			matrix[i][j], matrix[n-1-j][n-1-i] = matrix[n-1-j][n-1-i], matrix[i][j]
		}
	}

	for i := 0; i < n/2; i++ {
		for j := 0; j < n; j++ {
			matrix[i][j], matrix[n-1-i][j] = matrix[n-1-i][j], matrix[i][j]
		}
	}
}

File: leetcode/Arrays/leetcode-0048/48.java
class Solution {
    public void rotate(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return;
        }

        int row = matrix.length;
        int col = matrix.length;

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (i + j >= row - 1) {
                    continue;
                }
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[col - 1 - j][row - 1 - i];
                matrix[col - 1 - j][row - 1 - i] = tmp;
            }
        }
        for (int i = 0; i < row / 2; i++) {
            for (int j = 0; j < col; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[row - 1 - i][j];
                matrix[row - 1 - i][j] = tmp;
            }
        }
    }
}